### PR TITLE
Add support for function breakpoints

### PR DESCRIPTION
**Pull request by thegecko**
Signed-off-by: Rob Moran <rob.moran@arm.com>

Introduced support for adding function breakpoints, see:

https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_function-breakpoints

### CODE REVIEW

File: src/mi/breakpoint.ts [hunk 1]
```diff
@@ -66,3 +66,8 @@ export function sendBreakDelete(gdb: GDBBackend, request: {
 export function sendBreakList(gdb: GDBBackend): Promise<MIBreakListResponse> {
     return gdb.sendCommand('-break-list');
 }
+
+export function sendBreakFunctionInsert(gdb: GDBBackend, fn: string) {
+    const command = `-break-insert -t --function ${fn}`;
```
Comment (jonahgraham) [inline src/mi/breakpoint.ts]: Intentional temporary breakpoint?

Comment (thegecko) [reply src/mi/breakpoint.ts]: No, thanks

**Comment by thegecko** [comment]
Tests would be a lot easier to write if they worked on MacOS or ran on PRs (with the output shown)

**Comment by jonahgraham** [comment]
> Tests would be a lot easier to write if they worked on MacOS or ran on PRs (with the output shown)

Please add details on Issue #12 - or raise a new issue if needed.

**Comment by thegecko** [comment]
> I think this still doesn't meet the DAP spec

Still WiP

**Comment by thegecko** [comment]
Rebased and added tests, still need to resolve the issue with not removing them...

**Comment by jonahgraham** [comment]
> @thegecko pushed 1 commit.
>
>    88fcd1f Support removing function breakpoints

@thegecko does this mean it is ready for review? Let me know when/if it is and I will try to get it in.

**Comment by thegecko** [comment]
@jonahgraham This should be good to go. I've got a change which de-duplicates the breakpoint resolving in `setBreakpoints` and `setFunctionBreakpoints`, but felt that should be a PR after this one. Let me know if you'd prefer to see it added here, however.

**Comment by jonahgraham** [comment]
> Let me know if you'd prefer to see it added here, however.

Lets avoid the duplication in the first place - if you are ready to add it now, please do that. It will save me having to compare the setBreakpoints and setFunctionBreakpoints versions.

**Comment by thegecko** [comment]
@jonahgraham, I've introduced a common `resolveBreakpoints()` function for determining the breakpoints to add and delete across `setBreakpoints()` and `setFunctionBreakpoints()`.

I've introduced a few more tests to exercise this new common function and I've tested it all locally.

It's probably worth focussing your checks on the breakpoint handling overall as there is some churn in that area.

I've also added a few minor bug-fixes and comments.

File: src/GDBDebugSession.ts [hunk 2]
```diff
@@ -328,20 +339,83 @@ export class GDBDebugSession extends LoggingDebugSession {
...
+            const gdbbps = result.BreakpointTable.body.filter((gdbbp) => {
+                // Only function breakpoints
+                return this.functionBreakpoints.indexOf(gdbbp.number) > -1;
+            });
+
+            const { inserts, existing, deletes } = this.resolveBreakpoints(args.breakpoints, gdbbps,
+                (vsbp, gdbbp) => {
+
+                // Always invalidate hit conditions as they have a one-way mapping to gdb ignore and temporary
+                if (vsbp.hitCondition) {
```
Comment (jonahgraham) [inline src/GDBDebugSession.ts]: `hitCondition` and `condition` - VSCode [does not yet have](https://github.com/microsoft/vscode/issues/3646) UI for this so it seems acceptable to not implement support for now, but we are at some point going to need our own fork of the debug protocol spec to document our differences and extensions and this is one of them.

File: src/integration-tests/functionBreakpoints.spec.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                {
+                    name: 'sub',
+                },
+            ],
+        });
+        await dc.configurationDoneRequest();
+        const event = await dc.waitForEvent('stopped') as StoppedEvent;
+        expect(event.body.reason).to.eq('function breakpoint');
+        const scope = await getScopes(dc);
+        expect(scope.frame.line).to.eq(2);
```
Comment (jonahgraham) [inline src/integration-tests/functionBreakpoints.spec.ts]: There is already a helper that does this:

```
        dc.assertStoppedLocation('function breakpoint', {line: 2});
```

**Pull request merged**